### DISCUSSION TITLE
How to configure TMC5160 on MKS Tinybee?

### DISCUSSION THREAD

**Discussion by ashokmanker** [grblHAL/core#497] [General]
I am using MKS Tinybee board.  I know mks Tinybee don't support tmc5160 natively. But I have successfully configured tmc5160T on selected axis on FluidNC. After few modifications and jumper wiring I was able use tmc5160 with spi interface. 
How can I configure tmc5160 on selected axis on grblhal? I am using tmc5160 in Daisy chain config. 
Any help will be appreciated

**Comment by terjeio** [comment]
You'll have to modify the map file, [see this](https://github.com/grblHAL/ESP32/blob/master/main/boards/xPro_v5_map.h) for an example.

**Reply by ashokmanker** [reply]
I have seen this and configured but where do i define driver index since I have wired in daisy chain. In FluidNC i defined
Miso, mosi, clk, cs. Miso mosi clk same as sdcard. Defined separate cs for tmc. Then defined driver index for each axis which is using tmc according to position of driver in chain.

**Reply by terjeio** [reply]
Driver CS is defined [here](https://github.com/grblHAL/ESP32/blob/8147b87d62c9f9c09d9163d5a430c6cf8f36af70/main/boards/xPro_v5_map.h#L105), chain is X first, then Y, Z , M3 (A or ganged motor) etc, SPI pins shared with any SD card.

**Reply by ashokmanker** [reply]
I have done that. And defined Trinamic_spi_enable 1Trinamic_enable 5160Motor_cs_pin gpio_num_14Estop_enable 1Estop_pin auxinput0_pin
I want to use drv8825 on xyzAnd 5160 on a,b in daisychain
(In FluidNC i define spi_index 1 & 2 for a and b respectively)

**Reply by ashokmanker** [reply]
always this error comes. [MSG:Warning: Could not communicate with stepper driver!]

**Comment by ashokmanker** [comment]
I have done that. And defined Trinamic_spi_enable 1Trinamic_enable 5160Motor_cs_pin gpio_num_14Estop_enable 1Estop_pin auxinput0_pin
I want to use drv8825 on xyzAnd 5160 on a,b in daisychain
(In FluidNC i define spi_index 1 & 2 for a and b respectively)

Yahoo Mail: Search, organise, conquer 
 
  On Fri, 26 Apr 2024 at 15:10, Terje ***@***.***> wrote:   


Driver CS is defined here, chain is X first, then Y, Z , M3 (A or ganged motor) etc, SPI pins shared with any SD card.

**Reply by terjeio** [reply]
> I want to use drv8825 on xyzAnd 5160 on a,b in daisychain

Then you have to allow mixed drivers by setting [this symbol](https://github.com/grblHAL/ESP32/blob/8147b87d62c9f9c09d9163d5a430c6cf8f36af70/main/boards/xPro_v5_map.h#L38C1-L38C33) to `1` in your map file. This enables the [$338 setting](https://github.com/grblHAL/Plugins_motor/#settings), enable the A motor first and verify that communication is working before enabling B. At any time you can send `M122I` to reinitailize the driver(s), `M122` will report the current state.

**Reply by ashokmanker** [reply]
I have already tried that. After setting mixed mode to 1. Then there is setting appears in web config for tirnamic drivers to enable driver on axis. Tried m122i to reinitialise but failed.
Now testing single 5160 on b axis. Even tried defining trinamic_miso /mosi csm4_pin for b axis but everytime driver fails

**Reply by ashokmanker** [reply]
Is there any solution for working TMC 5160 drivers?
Tried hard but evertime driver fails.

**Reply by terjeio** [reply]
> Tried hard but evertime driver fails.

How? Crashes completely or with this message: `[MSG:Warning: Could not communicate with stepper driver!]`

**Reply by ashokmanker** [reply]
Yes. Could not communicate with stepper error. By the way I am compiling on linux with esp-idf. Because firmware compiled on platformio always causes CPU fault on ESP32.

**Reply by terjeio** [reply]
Have you commented out M3_LIMIT_PIN and M4_LIMIT_PIN in the map file? These pins are used for SPI and cannot be used as limit inputs.

**Reply by ashokmanker** [reply]
Yes i have already commented out those pins. And double checked for conflicting pin assignment

**Reply by terjeio** [reply]
Somewhat improved SPI code for the ESP32 was committed yesterday. Tested ok (no motor connected) with a single driver for the A-axis ($338=8) on a breadboarded ESP32 and with a single TMC5160.
I have not tested with the SD card enabled, not sure that will work since it has separate and possibly incompatible SPI initialization.
